### PR TITLE
PYTHON-3461 Test FaaS (AWS Lambda) Behavior Per Driver

### PR DESCRIPTION
**Pull request by blink1073**

### CODE REVIEW

File: .evergreen/config.yml
Comment (ShaneHarvey) [inline .evergreen/config.yml:3313]: Does this need to use rhel87-large? Or can we use rhel87-small instead?

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]
The [tests are failing with](https://spruce.mongodb.com/task/mongo_python_driver_rhel8_test_lambda_test_aws_lambda_deployed_patch_c6a6ea60663ce37cbd5128f1a83bc46e54fb2071_64b7f3f67742aeee93c6261e_23_07_19_14_32_23/logs?execution=0&sortBy=STATUS&sortDir=ASC):
```
 [2023/07/19 14:46:54.303] Lambda function ARN: arn:aws:lambda:us-east-1:579766882180:function:dbx-python-lambda-e1244092-MongoDBFunction-3wLFo93jPKfd
 [2023/07/19 14:46:54.303] ++ aws lambda invoke --function-name arn:aws:lambda:us-east-1:579766882180:function:dbx-python-lambda-e1244092-MongoDBFunction-3wLFo93jPKfd --log-type Tail lambda-invoke-standard.json
 [2023/07/19 14:47:09.209] {
 [2023/07/19 14:47:09.209]     "StatusCode": 200,
 [2023/07/19 14:47:09.209]     "FunctionError": "Unhandled",
 [2023/07/19 14:47:09.209]     "LogResult": "...",
 [2023/07/19 14:47:09.209]     "ExecutedVersion": "$LATEST"
 [2023/07/19 14:47:09.209] }
 [2023/07/19 14:47:09.306] {"errorMessage": "The \"dnspython\" module must be installed to use mongodb+srv:// URIs. To fix this error install pymongo again:\n /var/lang/bin/python3.9 -m pip install pymongo>=4.3", "errorType": "ConfigurationError", "requestId": "", "stackTrace": ["  File \"/var/lang/lib/python3.9/importlib/__init__.py\", line 127, in import_module\n    return _bootstrap._gcd_import(name[level:], package, level)\n", "  File \"<frozen importlib._bootstrap>\", line 1030, in _gcd_import\n", "  File \"<frozen importlib._bootstrap>\", line 1007, in _find_and_load\n", "  File \"<frozen importlib._bootstrap>\", line 986, in _find_and_load_unlocked\n", "  File \"<frozen importlib._bootstrap>\", line 680, in _load_unlocked\n", "  File \"<frozen importlib._bootstrap_external>\", line 850, in exec_module\n", "  File \"<frozen importlib._bootstrap>\", line 228, in _call_with_frames_removed\n", "  File \"/var/task/app.py\", line 99, in <module>\n    client = MongoClient(os.environ[\"MONGODB_URI\"], event_listeners=listeners)\n", "  File \"/var/task/pymongo/mongo_client.py\", line 748, in __init__\n    res = uri_parser.parse_uri(\n", "  File \"/var/task/pymongo/uri_parser.py\", line 470, in parse_uri\n    raise ConfigurationError(\n"]}
```

Could we make it so that this kind of error causes the EVG task to fail?

As for why this is happening, the dnspython package (and any other python dependencies) actually needs to be vendored into the lambda app package: https://docs.aws.amazon.com/lambda/latest/dg/python-package.html#python-package-dependencies

File: .evergreen/config.yml
Comment (blink1073) [reply .evergreen/config.yml:3313]: Small works

**Comment by blink1073** [comment]
Failing test: https://evergreen.mongodb.com/task_log_raw/mongo_python_driver_rhel8_test_lambda_test_aws_lambda_deployed_patch_1a249386cf704150369f08efc684b0c10686c1fc_64bfb150d1fe07b0370db59e_23_07_25_11_26_13/4?type=ALL

`Exiting due to FunctionError!`.

Adding `dnspython` to to the `requirements.txt`.

**Review by ShaneHarvey — APPROVED** [review]
LGTM but I think we should wait until your driver-tools fix is merged.

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]
Just noticed the tests are still failing:
```
[2023/07/25 16:49:46.758] ++ cat lambda_output.txt
[2023/07/25 16:49:46.759] {"errorMessage": "", "errorType": "AssertionError", "requestId": "", "stackTrace": ["  File \"/var/lang/lib/python3.9/importlib/__init__.py\", line 127, in import_module\n    return _bootstrap._gcd_import(name[level:], package, level)\n", "  File \"<frozen importlib._bootstrap>\", line 1030, in _gcd_import\n", "  File \"<frozen importlib._bootstrap>\", line 1007, in _find_and_load\n", "  File \"<frozen importlib._bootstrap>\", line 986, in _find_and_load_unlocked\n", "  File \"<frozen importlib._bootstrap>\", line 680, in _load_unlocked\n", "  File \"<frozen importlib._bootstrap_external>\", line 850, in exec_module\n", "  File \"<frozen importlib._bootstrap>\", line 228, in _call_with_frames_removed\n", "  File \"/var/task/app.py\", line 26, in <module>\n    assert has_bson_c()\n"]}{
[2023/07/25 16:49:46.759]     "StatusCode": 200,
[2023/07/25 16:49:46.759]     "FunctionError": "Unhandled",
```
https://spruce.mongodb.com/task/mongo_python_driver_rhel8_test_lambda_test_aws_lambda_deployed_patch_c6a6ea60663ce37cbd5128f1a83bc46e54fb2071_64bfdac42a60edb3ead68153_23_07_25_14_23_00/logs?execution=5

File: .evergreen/config.yml
Comment (blink1073) [reply .evergreen/config.yml:101]: Yes, and some clean up as well.  I agree we should wait for that PR to be merged

**Comment by blink1073** [comment]
> Just noticed the tests are still failing

Yes, I'm still working on installing the local version of pymongo vs the one from pypi.

**Comment by blink1073** [comment]
Moved to draft pending https://github.com/mongodb-labs/drivers-evergreen-tools/pull/331

**Review by ShaneHarvey — APPROVED** [review]
LGTM!

**Pull request merged**
